polar-core/src/kb.rs

```diff
@@ -1,26 +1,30 @@
 use std::collections::HashMap;
 
-use crate::error::PolarResult;
+use crate::error::ParameterError;
+use crate::error::{PolarError, PolarResult};
 
 pub use super::bindings::Bindings;
 use super::counter::Counter;
 use super::rules::*;
 use super::sources::*;
 use super::terms::*;
+use std::sync::Arc;
 
 /// A map of bindings: variable name → value. The VM uses a stack internally,
 /// but can translate to and from this type.
 
 #[derive(Default)]
 pub struct KnowledgeBase {
     pub constants: Bindings,
+    mro: HashMap<Symbol, TermList>,
 
     /// Map from loaded files to the source ID
     pub loaded_files: HashMap<String, u64>,
     /// Map from source code loaded to the filename it was loaded as
     pub loaded_content: HashMap<String, String>,
 
-    pub rules: HashMap<Symbol, GenericRule>,
+    rules: HashMap<Symbol, GenericRule>,
+    rule_templates: HashMap<Symbol, GenericRule>,
```

Comment:
The main reason for this is to reuse other code for parsing in rewriting rules. Since rule prototypes have the same structure as rules I think it makes sense.

Reply:
Why do rule prototypes need rewriting?

Reply:
It doesn't really look like you're using `GenericRule` at all. You  might want to just store this as `HashMap<Symbol, Vec<Rule>>` Or `HashMap<(Symbol, u64), Vec<Rule>>` if you want to include arity in the key.